enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,133 @@
+
+However, these features only work with a running cluster, users cannot configure desired RAID and BIOS during IPI deployments.
+
+### Goals
+
+- Allow users to configure RAID and BIOS in ***install-config.yaml***.
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. Add two new fields *raid* and *firmware* to *platform.baremetal.hosts* in the [install-config.yaml](https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml) file.
+2. Openshift Installer process the field contents into manifests.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for master nodes and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Add two fields
+
+New *platform.baremetal.hosts* fields called *raid* and *firmware*.
+
+```yaml
+platform:
+  baremetal:
+    ...
+    hosts:
+      - name: master-0
+        role: master
+        bmc:
+          address: IPAddress
+          username: UserName
+          password: PassWord
+        bootMACAddress: MACAddress
+        raid: RAIDConfig
+        firmware: BIOSConfig
+```
+
+RAID feature has been implemented in **BMO**, so the *raid* field here is the same as the [BMH](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/config/crd/bases/metal3.io_baremetalhosts.yaml).
+The *firmware* field is the same as the *spec.firmware* field in **BMH** which is been advancing by [#302](https://github.com/metal3-io/baremetal-operator/pull/302).
+
+#### Process the fields in installer
+
+For master nodes, call the `BuildTargetRAIDCfg` method in **BMO** to process the *raid* field into *target_raid_config*, and finally write *target_raid_config* into the ***terraform.baremetal.auto.tfvars.json*** file.
+
+For worker nodes, copy *raid* and *firmware* field to **BMH**.
+
+#### Process the fields in terraform-provider-ironic
+
+Add *target_raid_config* and *bios_settings* fields to terraform-provider-ironic API, 
+process the two fields by [manual cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#manual-cleaning).
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Test Plan
+
+- Unit tests for determining the configuration of RAID and BIOS passed to Ironic meeting expectations.
+- e2e tests for determining the configuration of RAID and BIOS configured during the IPI deployments.
```

Comment:
CI for OpenShift uses VMs to simulate bare metal. Do we have an emulator for testing the BIOS API? Can we test RAID configuration using VMs?

Reply:
> Do we have an emulator for testing the BIOS API?

No, currently we don't have an emulator for testing the BIOS API.

> Can we test RAID configuration using VMs?

Yes, we can use VMs to test software RAID.

Reply:
https://opendev.org/openstack/sushy-tools has rudimentary BIOS support, we can use it for some basic testing probably.

Reply:
> https://opendev.org/openstack/sushy-tools has rudimentary BIOS support, we can use it for some basic testing probably.

@dtantsur Thanks for your advice.